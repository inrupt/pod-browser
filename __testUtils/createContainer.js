/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  createSolidDataset,
  createThing,
  addDatetime,
  addDecimal,
  addInteger,
  setThing,
} from "@inrupt/solid-client";
import { rdf, ldp } from "rdf-namespaces";

export const TIMESTAMP = new Date(Date.UTC(2020, 5, 2, 15, 59, 21));

export default function createContainer(
  url = "https://user.dev.inrupt.net/",
  type = ldp.BasicContainer
) {
  const thing = createThing({ url });
  const publicContainer = [
    (t) => addUrl(t, rdf.type, type),
    (t) => addUrl(t, rdf.type, ldp.Container),
    (t) => addDatetime(t, "http://purl.org/dc/terms/modified", TIMESTAMP),
    (t) =>
      addDecimal(t, "http://www.w3.org/ns/posix/stat#mtime", 1591131561.195),
    (t) => addInteger(t, "http://www.w3.org/ns/posix/stat#size", 4096),
    (t) => addUrl(t, ldp.contains, "https://user.dev.inrupt.net/public/games/"),
  ].reduce((acc, fn) => fn(acc), thing);

  return setThing(createSolidDataset(), publicContainer);
}
