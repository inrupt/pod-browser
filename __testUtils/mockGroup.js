/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  getStringNoLocale,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/group";
import { vcardExtras } from "../src/models/addressBook";

function mockGroupThing(name, groupThingUrl) {
  return chain(
    mockThingFrom(groupThingUrl),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, vcard.fn, name)
  );
}

function mockIndexThing(addressBook, groupThingUrl) {
  return setUrl(addressBook.thing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
  );
}

// export function addGroupsToAddressBook(addressBook, groupsToAdd) {
//   const { groups, ...rest } = addressBook;
//   return {
//     ...rest,
//     groups: {
//       iri: addressBook.groups.iri,
//       dataset: chain(
//         addressBook.groups.dataset,
//         ...groupsToAdd.map(({ dataset, iri }) => (d) => {
//           const group = getThing(dataset, iri);
//           const name = getStringNoLocale(group, vcard.fn);
//           return setThing(d, mockGroupThing(name, iri));
//         }),
//         ...groupsToAdd.map((group) => (d) =>
//           setThing(d, mockIndexThing(addressBook, group.iri))
//         )
//       ),
//     },
//   };
// }

export default function mockGroup(
  addressBook,
  name,
  { url, id, members } = {}
) {
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = chain(
    mockGroupThing(name, groupThingUrl),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    )
  );
  return {
    dataset: chain(
      mockSolidDatasetFrom(groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
