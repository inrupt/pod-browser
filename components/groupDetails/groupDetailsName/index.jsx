/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint react/forbid-prop-types: off */

import React, { useContext, useEffect, useState } from "react";
import { useSession } from "@inrupt/solid-ui-react";
import {
  Button,
  Form,
  InputGroup,
  SimpleInput,
} from "@inrupt/prism-react-components";
import { makeStyles } from "@material-ui/styles";
import { createStyles } from "@material-ui/core";
import { useBem } from "@solid/lit-prism-patterns";
import { getGroupName } from "../../../src/models/group";
import styles from "./styles";
import Spinner from "../../spinner";
import { renameGroup } from "../../../src/models/contact/group";
import ErrorMessage from "../../errorMessage";
import { getContactAllFromContactsIndex } from "../../../src/models/contact";
import GroupAllContext from "../../../src/contexts/groupAllContext";
import AddressBookContext from "../../../src/contexts/addressBookContext";
import GroupContext from "../../../src/contexts/groupContext";

export const TESTCAFE_ID_GROUP_DETAILS_NAME = "group-details-name";
export const TESTCAFE_ID_GROUP_DETAILS_NAME_FIELD = "group-details-name-field";
export const TESTCAFE_ID_GROUP_DETAILS_NAME_BUTTON =
  "group-details-name-button";

export const MESSAGE_GROUP_DETAILS_NAME_REQUIRED = "Name is required";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export default function GroupDetailsName() {
  const { data: group, error: groupError, mutate: mutateGroup } = useContext(
    GroupContext
  );
  const [processing, setProcessing] = useState(false);
  const bem = useBem(useStyles());
  const [groupName, setGroupName] = useState("");
  const fieldId = "GroupName";
  const { data: addressBook, error: addressBookError } = useContext(
    AddressBookContext
  );
  const { fetch } = useSession();
  const { error: groupsError, mutate: mutateGroups } = useContext(
    GroupAllContext
  );

  useEffect(() => {
    if (!group) return;
    setGroupName(getGroupName(group));
  }, [group]);

  const onSubmit = async (event) => {
    event.preventDefault();
    if (!groupName) return;
    setProcessing(true);

    const { group: updatedGroup, groupIndex } = await renameGroup(
      addressBook,
      group,
      groupName,
      fetch
    );
    const updatedGroups = getContactAllFromContactsIndex(groupIndex);
    await Promise.all([mutateGroup(updatedGroup), mutateGroups(updatedGroups)]);
    setProcessing(false);
  };

  const loading = processing || !group;

  if (loading) return <Spinner />;

  const error = addressBookError || groupError || groupsError;

  if (error) return <ErrorMessage error={error} />;

  return (
    <Form
      onSubmit={onSubmit}
      className={bem("group-details-name")}
      data-testid={TESTCAFE_ID_GROUP_DETAILS_NAME}
    >
      <InputGroup
        id={fieldId}
        label="Group Name"
        variant="with-button"
        invalidMessage={groupName ? null : MESSAGE_GROUP_DETAILS_NAME_REQUIRED}
      >
        <SimpleInput
          id={fieldId}
          value={groupName}
          onChange={(event) => setGroupName(event.target.value)}
          variant="with-button"
          data-testid={TESTCAFE_ID_GROUP_DETAILS_NAME_FIELD}
          required
        />
        <Button
          variant="with-input"
          type="submit"
          data-testid={TESTCAFE_ID_GROUP_DETAILS_NAME_BUTTON}
        >
          Save
        </Button>
      </InputGroup>
    </Form>
  );
}
