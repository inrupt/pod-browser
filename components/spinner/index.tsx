import React, { ReactElement } from "react";
import { CircularProgress, Container } from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles(() => ({
  "MuiCicularProgress-root": {
    display: "block",
    position: "fixed",
    top: "50%",
    right: "50%",
  },
}));

export default function Spinner(): ReactElement {
  const classes = useStyles();
  return (
    <Container>
      <CircularProgress
        className={classes["MuiCicularProgress-root"]}
        color="primary"
      />
    </Container>
  );
}
