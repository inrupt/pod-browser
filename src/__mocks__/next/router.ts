export default {
  push: jest.fn().mockResolvedValue(null),
};
