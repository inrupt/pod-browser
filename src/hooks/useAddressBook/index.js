/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useContext, useEffect, useState } from "react";
import { fetchLitDataset } from "@inrupt/solid-client";
import SessionContext from "../../contexts/sessionContext";
import useAuthenticatedProfile from "../useAuthenticatedProfile";
import { contactsContainerIri, saveNewAddressBook } from "../../addressBook";
import { getResource } from "../../solidClientHelpers/resource";
import { ERROR_CODES, isHTTPError } from "../../solidClientHelpers/utils";

export default function useAddressBook() {
  const [addressBook, setAddressBook] = useState(null);
  const [error, setError] = useState(null);
  const { session } = useContext(SessionContext);
  const profile = useAuthenticatedProfile();

  useEffect(() => {
    if (!session.info.isLoggedIn || !profile) return;
    const { pods, webId } = profile;
    const { fetch } = session;
    const contactsIri = contactsContainerIri(pods[0]);

    (async () => {
      let addressBookUri;

      const {
        response: existingAddressBook,
        error: existingError,
      } = await getResource(contactsIri, fetch);

      if (existingAddressBook) {
        setAddressBook(existingAddressBook.dataset);
        return;
      }

      if (existingError && !isHTTPError(existingError, ERROR_CODES.NOT_FOUND)) {
        const {
          response: newAddressBook,
          error: newError,
        } = await saveNewAddressBook(
          {
            iri: contactsIri,
            owner: webId,
          },
          fetch
        );
        if (newError) {
          setError(newError);
          return;
        }
        addressBookUri = newAddressBook.iri;
      } else if (existingError) {
        setError(existingError);
        return;
      }

      setAddressBook(await fetchLitDataset(addressBookUri, { fetch }));
    })();
  }, [session, profile]);

  return [addressBook, error];
}
