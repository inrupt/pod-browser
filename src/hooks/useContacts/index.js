/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useSession } from "@inrupt/solid-ui-react";
import useSWR from "swr";
import { thingAsMarkdown } from "@inrupt/solid-client";
import useAddressBook from "../useAddressBook";
import { getContactAll } from "../../models/contact";
import { compareArray } from "../../models/array";

function reduceToString(array) {
  return array.reduce(
    (memo, item) => memo.concat(thingAsMarkdown(item.thing)),
    ""
  );
}

export default function useContacts(type) {
  const { fetch } = useSession();
  const { data: addressBook, error: addressBookError } = useAddressBook();
  return useSWR(
    ["contacts", addressBook, type],
    async () => {
      if (!addressBook && !addressBookError) return null;
      if (addressBookError) throw addressBookError;
      return getContactAll(addressBook, [type], fetch);
    },
    {
      // revalidateOnMount: true,
      // revalidateOnFocus: false,
      // revalidateOnReconnect: false,
      // refreshInterval: 0,
      shouldRetryOnError: false,
      // errorRetryCount: 0,
      compare: (a, b) =>
        compareArray(a, b, (c, d) => reduceToString(c) === reduceToString(d)),
      // compare: (a, b) => {
      //   if (typeof a === "undefined") {
      //     return typeof b === "undefined";
      //   }
      //   if (typeof b === "undefined") {
      //     return typeof a === "undefined";
      //   }
      //   if (!Array.isArray(a) || !Array.isArray(b)) {
      //     return false;
      //   }
      //   if (a.length !== b.length) {
      //     return false;
      //   }
      //   const reduce = reduceToString(a) === reduceToString(b);
      //   // if (!reduce) {
      //   //   console.log("COMPARE ARRAY A", reduceToString(a));
      //   //   console.log("COMPARE ARRAY B", reduceToString(b));
      //   // }
      //   return reduce;
      // },
    }
  );
}
