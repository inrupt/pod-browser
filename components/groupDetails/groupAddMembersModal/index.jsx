/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useContext, useState } from "react";
import {
  Button,
  FormModal,
  Modal,
  ModalBody,
  ModalButtonsContainer,
  ModalContainer,
  ModalTitle,
} from "@inrupt/prism-react-components";
import T from "prop-types";
import { useSession } from "@inrupt/solid-ui-react";
import AgentMultiSelectPicker from "../../agentMultiSelectPicker";
import useContacts from "../../../src/hooks/useContacts";
import {
  PERSON_CONTACT,
  savePersonWithSchema,
} from "../../../src/models/contact/person";
import { GROUP_CONTACT } from "../../../src/models/contact/group";
import GroupContext from "../../../src/contexts/groupContext";
import Spinner from "../../spinner";
import {
  getGroupMemberUrlAll,
  getGroupUrl,
  updateGroupMembers,
} from "../../../src/models/group";
import { TEMP_CONTACT } from "../../../src/models/contact/temp";
import AddressBookContext from "../../../src/contexts/addressBookContext";
import { getOriginalUrlForContactAll } from "../../../src/models/contact";

export default function GroupAddMembersModal({ open, handleClose }) {
  const { fetch } = useSession();
  const { data: addressBook } = useContext(AddressBookContext);
  const { data: group, mutate: mutateGroup, isValidating } = useContext(
    GroupContext
  );
  const { data: contacts, mutate: mutateContacts } = useContacts(
    [PERSON_CONTACT, GROUP_CONTACT],
    {
      revalidateOnFocus: false,
      errorRetryCount: 0,
    }
  );
  const [selectedChanged, setSelectedChanged] = useState({});
  const [unselectedChanged, setUnselectedChanged] = useState({});
  const [processing, setProcessing] = useState(false);

  if (isValidating) return <Spinner />;

  const disabled = [getGroupUrl(group)];
  const selected = getGroupMemberUrlAll(group);

  const handleChange = (selectedChange, unselectedChange) => {
    setSelectedChanged(selectedChange);
    setUnselectedChanged(unselectedChange);
  };

  const handleSubmit = async () => {
    setProcessing(true);
    const newContacts = Object.values(selectedChanged).filter((model) =>
      TEMP_CONTACT.isOfType(model.thing)
    );
    if (newContacts.length) {
      // handle adding new contact(s)
      const agentOriginalUrlAll = await getOriginalUrlForContactAll(
        addressBook,
        [PERSON_CONTACT],
        fetch
      );
      const actualNewContacts = newContacts.filter(
        (tempAgent) =>
          agentOriginalUrlAll.indexOf(
            tempAgent.type.getOriginalUrl(tempAgent)
          ) === -1
      );
      await Promise.all(
        actualNewContacts.map((tempAgent) => {
          const contactSchema = {
            webId: tempAgent.type.getOriginalUrl(tempAgent),
            fn: tempAgent.type.getName(tempAgent),
          };
          return savePersonWithSchema(addressBook, contactSchema, fetch);
        })
      );
      await mutateContacts();
    }
    // update group members
    const membersToAdd = Object.keys(selectedChanged);
    const membersToRemove = Object.keys(unselectedChanged);
    const updatedGroup = await updateGroupMembers(
      group,
      membersToAdd,
      membersToRemove,
      fetch
    );
    await mutateGroup(updatedGroup);
    setProcessing(false);
    handleClose();
  };

  return (
    <Modal open={open} onClose={handleClose}>
      <ModalContainer>
        <ModalTitle>Edit Group</ModalTitle>
        <ModalBody>
          {processing && <Spinner />}
          {!processing && (
            <AgentMultiSelectPicker
              contacts={contacts}
              onChange={handleChange}
              selected={selected}
              disabled={disabled}
            />
          )}
        </ModalBody>
        {!processing && (
          <ModalButtonsContainer>
            <Button variant="secondary" onClick={handleClose}>
              Cancel
            </Button>
            <Button onClick={handleSubmit}>Save Group</Button>
          </ModalButtonsContainer>
        )}
      </ModalContainer>
    </Modal>
  );
}

GroupAddMembersModal.propTypes = {
  handleClose: T.func.isRequired,
  open: T.bool.isRequired,
};
