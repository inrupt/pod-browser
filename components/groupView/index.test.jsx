/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useRouter } from "next/router";
import { renderWithTheme } from "../../__testUtils/withTheme";
import useContacts from "../../src/hooks/useContacts";
import { TESTCAFE_ID_SPINNER } from "../spinner";
import GroupView from "./index";
import { TESTCAFE_ID_GROUP_VIEW_EMPTY } from "./groupViewEmpty";
import { TESTCAFE_ID_ERROR_MESSAGE } from "../errorMessage";
import useAddressBook from "../../src/hooks/useAddressBook";
import useGroup from "../../src/hooks/useGroup";
import mockGroup from "../../__testUtils/mockGroup";
import renderGroupsPage from "../../__testUtils/renderGroupsPage";
import { TESTCAFE_ID_GROUP_DETAILS } from "../groupDetails";

jest.mock("../../src/hooks/useAddressBook");
const mockedAddressBookHook = useAddressBook;

jest.mock("../../src/hooks/useContacts");
const mockedContactsHook = useContacts;

jest.mock("../../src/hooks/useGroup");
const mockedGroupHook = useGroup;

jest.mock("next/router");
const mockedRouterHook = useRouter;

const group1Name = "Group 1";
const group1Url = "http://example.com/group1.ttl#this";
const group1 = mockGroup(group1Name, group1Url);

describe("GroupView", () => {
  beforeEach(() => {
    mockedAddressBookHook.mockReturnValue({});
    mockedContactsHook.mockReturnValue({ data: [group1] });
    mockedGroupHook.mockReturnValue({ data: group1 });
    mockedRouterHook.mockReturnValue({ query: { iri: group1Url } });
  });

  it("renders group details", () => {
    const { getByTestId } = renderGroupsPage(<GroupView />);
    expect(getByTestId(TESTCAFE_ID_GROUP_DETAILS)).toBeDefined();
  });

  it("renders an empty list when no groups are loaded", () => {
    mockedRouterHook.mockReturnValue({ query: {} });
    mockedContactsHook.mockReturnValue({ data: [] });
    mockedGroupHook.mockReturnValue({ data: null });
    const { getByTestId } = renderGroupsPage(<GroupView />);
    expect(getByTestId(TESTCAFE_ID_GROUP_VIEW_EMPTY)).toBeDefined();
  });

  it("renders a spinner while groups are loading", () => {
    mockedContactsHook.mockReturnValue({});
    const { getByTestId } = renderGroupsPage(<GroupView />);
    expect(getByTestId(TESTCAFE_ID_SPINNER)).toBeDefined();
  });

  it("renders a spinner while group is loading", () => {
    mockedGroupHook.mockReturnValue({});
    const { getByTestId } = renderGroupsPage(<GroupView />);
    expect(getByTestId(TESTCAFE_ID_SPINNER)).toBeDefined();
  });

  it("renders an error if something goes wrong when loading groups", () => {
    const errorMessage = "error";
    mockedContactsHook.mockReturnValue({ error: new Error(errorMessage) });
    const { getByTestId } = renderGroupsPage(<GroupView />);
    expect(getByTestId(TESTCAFE_ID_ERROR_MESSAGE).innerHTML).toContain(
      errorMessage
    );
  });
});
