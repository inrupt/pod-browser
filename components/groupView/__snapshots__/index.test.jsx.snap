// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupView renders an empty list when no groups are loaded 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-group-view"
    data-testid="group-view"
  >
    <form
      class="PodBrowser-form PodBrowser-group-details-name"
    >
      <div
        class="PodBrowser-input-group"
      >
        <label
          class="PodBrowser-label"
          for="GroupName"
        >
          Group Name
        </label>
        <div
          class="PodBrowser-input-group__inner PodBrowser-input-group__inner--with-button"
        >
          <input
            class="PodBrowser-input"
            data-testid="group-name-field"
            id="GroupName"
            type="text"
            value="Group 1"
          />
          <button
            class="PodBrowser-button PodBrowser-button--with-input"
            type="submit"
          >
            <span
              class="PodBrowser-button__text"
            >
              Save
            </span>
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
